{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['L161Q']\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "from codon_table import dna_codons\n",
    "from colors import rasmol\n",
    "\n",
    "lpla_seq = \"AGCACCCTGAGACTGCTGATCAGCGACAGCTACGACCCCTGGTTCAACCTGGCCGTGGAGGAGTGCATCTTCAGACAGATGCCCGCCACCCAGAGAGTGCTGTTCCTGGTGAGAAACGCCGACACCGTGGTGATCGGCAGAGCCCAGAACCCCTGGAAGGAGTGCAACACCAGAAGAATGGAGGAGGACAACGTGAGACTGGCCAGAAGAAGCAGCGGCGGCGGCGCCGTGTTCCACGACCTGGGCAACACCTGCTTCACCTTCATGGCCGGCAAGCCCGAGTACGACAAGACCATCAGCACCAGCATCGTGCTGAACGCCCTGAACGCCCTGGGCGTGAGCGCCGAGGCCAGCGGCAGAAACGACCTGGTGGTGAAGACCGTGGAGGGCGACAGAAAGGTGAGCGGCAGCGCCTACAGAGAGACCAAGGACAGAGGCTTCCACCACGGCACCCTGCTGCTGAACGCCGACCTGAGCAGACTGGCCAACTACCTGAACCCCGACAAGAAGAAGCTGGCCGCCAAGGGCATCACCAGCGTGAGAAGCAGAGTGACCAACCTGACCGAGCTGCTGCCCGGCATCACCCACGAGCAGGTGTGCGAGGCCATCACCGAGGCCTTCTTCGCCCACTACGGCGAGAGAGTGGAGGCCGAGATCATCAGCCCCAACAAGACCCCCGACCTGCCCAACTTCGCCGAGACCTTCGCCAGACAGAGCAGCTGGGAGTGGAACTTCGGCCAGGCCCCCGCCTTCAGCCACCTGCTGGACGAGAGATTCACCTGGGGCGGCGTGGAGCTGCACTTCGACGTGGAGAAGGGCCACATCACCAGAGCCCAGGTGTTCACCGACAGCCTGAACCCCGCCCCCCTGGAGGCCCTGGCCGGCAGACTGCAGGGCTGCCTGTACAGAGCCGACATGCTGCAGCAGGAGTGCGAGGCCCTGCTGGTGGACTTCCCCGAGCAGGAGAAGGAGCTGAGAGAGCTGAGCGCCTGGATGGCCGGCGCCGTGAGATAAgaattctgcagatatccatcacactgg\"\n",
    "\n",
    "def translate(seq, codon_table):\n",
    "    protein = \"\"\n",
    "    for codon_idx in range(len(seq) // 3):\n",
    "        codon = seq[codon_idx * 3 : (codon_idx + 1) * 3].upper()\n",
    "        if codon in codon_table:\n",
    "            if codon_table[codon] == \"*\":\n",
    "                return protein\n",
    "            else:\n",
    "                protein += codon_table[codon]\n",
    "\n",
    "    return protein\n",
    "\n",
    "\n",
    "def find_mutations(wt_pseq, mut_pseq):\n",
    "    mutations = []\n",
    "    for a, b, idx in zip(wt_pseq.upper(), mut_pseq.upper(), range(len(wt_pseq))):\n",
    "        prot_idx = idx + 1\n",
    "        if a != b:\n",
    "            mutations.append(f\"{a}{prot_idx}{b}\")\n",
    "\n",
    "    return mutations\n",
    "\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    complement = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\"}\n",
    "    return \"\".join([complement[base.upper()] for base in seq[::-1]])\n",
    "\n",
    "\n",
    "match_score = 2\n",
    "mismatch_penalty = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"primers.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a447e7c61d44dbba5ae37d1fd201ad1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/234 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from pandas import isna\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "\n",
    "df = pd.read_csv(\"primers.csv\")\n",
    "\n",
    "lpla_seq = lpla_seq.upper()\n",
    "\n",
    "\n",
    "muts = []\n",
    "for primer in tqdm(df[\"Sequence\"]):\n",
    "\n",
    "    if pd.isna(primer):\n",
    "        muts.append(\"\")\n",
    "    else:\n",
    "        primer = primer.upper()\n",
    "        alignments = pairwise2.align.globalms(\n",
    "            lpla_seq, primer, match_score, mismatch_penalty, -10, -1, penalize_end_gaps=False\n",
    "        )\n",
    "        alignment = alignments[0]\n",
    "        mismatches = math.ceil(\n",
    "            (len(primer) * match_score - alignment.score) / (match_score - mismatch_penalty)\n",
    "        )\n",
    "\n",
    "        if mismatches > 3:\n",
    "            primer = reverse_complement(primer)\n",
    "            alignments = pairwise2.align.globalms(\n",
    "                lpla_seq,\n",
    "                primer,\n",
    "                match_score,\n",
    "                mismatch_penalty,\n",
    "                -10,\n",
    "                -1,\n",
    "                penalize_end_gaps=False,\n",
    "            )\n",
    "            alignment = alignments[0]\n",
    "            mismatches = math.ceil(\n",
    "                (len(primer) * match_score - alignment.score) / (match_score - mismatch_penalty)\n",
    "            )\n",
    "            if mismatches > 3:\n",
    "                raise ValueError(\n",
    "                    \"Primer does not match sequence! There are more than three mismatches in either orientation.\"\n",
    "                )\n",
    "\n",
    "        template_seq = alignment.seqA\n",
    "        primer_seq = alignment.seqB\n",
    "\n",
    "        merged_seq = \"\"\n",
    "        for ba, bb in zip(template_seq, primer_seq):\n",
    "            if bb == \"-\":\n",
    "                merged_seq += ba\n",
    "            else:\n",
    "                merged_seq += bb\n",
    "\n",
    "        original_pseq = translate(lpla_seq, dna_codons)\n",
    "        mutant_pseq = translate(merged_seq, dna_codons)\n",
    "        mutation = find_mutations(original_pseq, mutant_pseq)\n",
    "        \n",
    "        muts.append(\"\".join(mutation))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments = pairwise2.align.globalms(\n",
    "    lpla_seq.upper(),\n",
    "    \"GCGCCTGGATGGCCAAAGCCGTGAGATAAGAATTC\",\n",
    "    match_score,\n",
    "    mismatch_penalty,\n",
    "    -10,\n",
    "    -1,\n",
    "    penalize_end_gaps=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Mutation</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Pred_Mutation</th>\n",
       "      <th>Mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y139V</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC</td>\n",
       "      <td>Y139V</td>\n",
       "      <td>Y139V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTCACGGCGCTGCCGCTCACC</td>\n",
       "      <td>Y139V</td>\n",
       "      <td>Y139V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y139T</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCACCAGAGAGACCAAGGAC</td>\n",
       "      <td>Y139T</td>\n",
       "      <td>Y139T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTGGTGGCGCTGCCGCTCACC</td>\n",
       "      <td>Y139T</td>\n",
       "      <td>Y139T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y139S</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCAGCAGAGAGACCAAGGAC</td>\n",
       "      <td>Y139S</td>\n",
       "      <td>Y139S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>TCTGGTGTTGCAGGCCTTCCAGGGGTTCTGG</td>\n",
       "      <td>E54A</td>\n",
       "      <td>E54A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>E54H</td>\n",
       "      <td>F</td>\n",
       "      <td>CCCCTGGAAGCACTGCAACACCAGAAGAATG</td>\n",
       "      <td>E54H</td>\n",
       "      <td>E54H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTGGTGTTGCAGTGCTTCCAGGGGTTCTGGG</td>\n",
       "      <td>E54H</td>\n",
       "      <td>E54H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>A248P</td>\n",
       "      <td>F</td>\n",
       "      <td>AACTTCGGCCAGCCCCCCGCCTTCAGC</td>\n",
       "      <td>A248P</td>\n",
       "      <td>A248P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>TGAAGGCGGGGGGCTGGCCGAAGTTCC</td>\n",
       "      <td>A248P</td>\n",
       "      <td>A248P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>234 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Target_Mutation Orientation                         Sequence  \\\n",
       "0             Y139V           F  GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC   \n",
       "1               NaN           R  CTTGGTCTCTCTCACGGCGCTGCCGCTCACC   \n",
       "2             Y139T           F  GAGCGGCAGCGCCACCAGAGAGACCAAGGAC   \n",
       "3               NaN           R  CTTGGTCTCTCTGGTGGCGCTGCCGCTCACC   \n",
       "4             Y139S           F  GAGCGGCAGCGCCAGCAGAGAGACCAAGGAC   \n",
       "..              ...         ...                              ...   \n",
       "229             NaN           R  TCTGGTGTTGCAGGCCTTCCAGGGGTTCTGG   \n",
       "230            E54H           F  CCCCTGGAAGCACTGCAACACCAGAAGAATG   \n",
       "231             NaN           R  CTGGTGTTGCAGTGCTTCCAGGGGTTCTGGG   \n",
       "232           A248P           F      AACTTCGGCCAGCCCCCCGCCTTCAGC   \n",
       "233             NaN           R      TGAAGGCGGGGGGCTGGCCGAAGTTCC   \n",
       "\n",
       "    Pred_Mutation Mutation  \n",
       "0           Y139V    Y139V  \n",
       "1           Y139V    Y139V  \n",
       "2           Y139T    Y139T  \n",
       "3           Y139T    Y139T  \n",
       "4           Y139S    Y139S  \n",
       "..            ...      ...  \n",
       "229          E54A     E54A  \n",
       "230          E54H     E54H  \n",
       "231          E54H     E54H  \n",
       "232         A248P    A248P  \n",
       "233         A248P    A248P  \n",
       "\n",
       "[234 rows x 5 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "prev = \"\"\n",
    "real_targets = []\n",
    "for target in df['Target_Mutation']:\n",
    "    if pd.isna(target):\n",
    "        real_targets.append(prev)\n",
    "    else:\n",
    "        real_targets.append(target)\n",
    "        prev = target\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Target_Mutation'] = real_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('primers_checked.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Mutation</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Pred_Mutation</th>\n",
       "      <th>Primer_Correct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y139V</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC</td>\n",
       "      <td>Y139V</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y139V</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTCACGGCGCTGCCGCTCACC</td>\n",
       "      <td>Y139V</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y139T</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCACCAGAGAGACCAAGGAC</td>\n",
       "      <td>Y139T</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y139T</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTGGTGGCGCTGCCGCTCACC</td>\n",
       "      <td>Y139T</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y139S</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCAGCAGAGAGACCAAGGAC</td>\n",
       "      <td>Y139S</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>E54A</td>\n",
       "      <td>R</td>\n",
       "      <td>TCTGGTGTTGCAGGCCTTCCAGGGGTTCTGG</td>\n",
       "      <td>E54A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>E54H</td>\n",
       "      <td>F</td>\n",
       "      <td>CCCCTGGAAGCACTGCAACACCAGAAGAATG</td>\n",
       "      <td>E54H</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>E54H</td>\n",
       "      <td>R</td>\n",
       "      <td>CTGGTGTTGCAGTGCTTCCAGGGGTTCTGGG</td>\n",
       "      <td>E54H</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>A248P</td>\n",
       "      <td>F</td>\n",
       "      <td>AACTTCGGCCAGCCCCCCGCCTTCAGC</td>\n",
       "      <td>A248P</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>A248P</td>\n",
       "      <td>R</td>\n",
       "      <td>TGAAGGCGGGGGGCTGGCCGAAGTTCC</td>\n",
       "      <td>A248P</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>234 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Target_Mutation Orientation                         Sequence  \\\n",
       "0             Y139V           F  GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC   \n",
       "1             Y139V           R  CTTGGTCTCTCTCACGGCGCTGCCGCTCACC   \n",
       "2             Y139T           F  GAGCGGCAGCGCCACCAGAGAGACCAAGGAC   \n",
       "3             Y139T           R  CTTGGTCTCTCTGGTGGCGCTGCCGCTCACC   \n",
       "4             Y139S           F  GAGCGGCAGCGCCAGCAGAGAGACCAAGGAC   \n",
       "..              ...         ...                              ...   \n",
       "229            E54A           R  TCTGGTGTTGCAGGCCTTCCAGGGGTTCTGG   \n",
       "230            E54H           F  CCCCTGGAAGCACTGCAACACCAGAAGAATG   \n",
       "231            E54H           R  CTGGTGTTGCAGTGCTTCCAGGGGTTCTGGG   \n",
       "232           A248P           F      AACTTCGGCCAGCCCCCCGCCTTCAGC   \n",
       "233           A248P           R      TGAAGGCGGGGGGCTGGCCGAAGTTCC   \n",
       "\n",
       "    Pred_Mutation Primer_Correct  \n",
       "0           Y139V           True  \n",
       "1           Y139V           True  \n",
       "2           Y139T           True  \n",
       "3           Y139T           True  \n",
       "4           Y139S           True  \n",
       "..            ...            ...  \n",
       "229          E54A           True  \n",
       "230          E54H           True  \n",
       "231          E54H           True  \n",
       "232         A248P           True  \n",
       "233         A248P           True  \n",
       "\n",
       "[234 rows x 5 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"Pred_Mutation\"] = muts\n",
    "df[\"Primer_Correct\"] = [\n",
    "    pred == target if pred != \"\" else \"\"\n",
    "    for pred, target in zip(df[\"Pred_Mutation\"], df[\"Target_Mutation\"])\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Pred_Mutation'] = muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(\"GCGCCTGGATGGCCAAAGCCGTGAGATAAGAATTC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64.0"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments[0].score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCACCCTGAGACTGCTGATCAGCGACAGCTACGACCCCTGGTTCAACCTGGCCGTGGAGGAGTGCATCTTCAGACAGATGCCCGCCACCCAGAGAGTGCTGTTCCTGGTGAGAAACGCCGACACCGTGGTGATCGGCAGAGCCCAGAACCCCTGGAAGGAGTGCAACACCAGAAGAATGGAGGAGGACAACGTGAGACTGGCCAGAAGAAGCAGCGGCGGCGGCGCCGTGTTCCACGACCTGGGCAACACCTGCTTCACCTTCATGGCCGGCAAGCCCGAGTACGACAAGACCATCAGCACCAGCATCGTGCTGAACGCCCTGAACGCCCTGGGCGTGAGCGCCGAGGCCAGCGGCAGAAACGACCTGGTGGTGAAGACCGTGGAGGGCGACAGAAAGGTGAGCGGCAGCGCCTACAGAGAGACCAAGGACAGAGGCTTCCACCACGGCACCCTGCTGCTGAACGCCGACCTGAGCAGACTGGCCAACTACCTGAACCCCGACAAGAAGAAGCTGGCCGCCAAGGGCATCACCAGCGTGAGAAGCAGAGTGACCAACCTGACCGAGCTGCTGCCCGGCATCACCCACGAGCAGGTGTGCGAGGCCATCACCGAGGCCTTCTTCGCCCACTACGGCGAGAGAGTGGAGGCCGAGATCATCAGCCCCAACAAGACCCCCGACCTGCCCAACTTCGCCGAGACCTTCGCCAGACAGAGCAGCTGGGAGTGGAACTTCGGCCAGGCCCCCGCCTTCAGCCACCTGCTGGACGAGAGATTCACCTGGGGCGGCGTGGAGCTGCACTTCGACGTGGAGAAGGGCCACATCACCAGAGCCCAGGTGTTCACCGACAGCCTGAACCCCGCCCCCCTGGAGGCCCTGGCCGGCAGACTGCAGGGCTGCCTGTACAGAGCCGACATGCTGCAGCAGGAGTGCGAGGCCCTGCTGGTGGACTTCCCCGAGCAGGAGAAGGAGCTGAGAGAGCTGAGCGCCTGGATGGCCGGCGCCGTGAGATAAgaattctgcagatatccatcacactgg\n",
      "-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------GCGCCTGGATGGCCAAAGCCGTGAGATAAGAATTC---------------------\n"
     ]
    }
   ],
   "source": [
    "print(alignments[0].seqA)\n",
    "print(alignments[0].seqB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCGCCTGGATGGCCAAAGCCGTGAGATAAGAATTC'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reverse_complement(primer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_Mutation</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y139V</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTCACGGCGCTGCCGCTCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y139T</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCACCAGAGAGACCAAGGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTGGTGGCGCTGCCGCTCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y139S</td>\n",
       "      <td>F</td>\n",
       "      <td>GAGCGGCAGCGCCAGCAGAGAGACCAAGGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTTGGTCTCTCTGCTGGCGCTGCCGCTCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A48D</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>E265V</td>\n",
       "      <td>F</td>\n",
       "      <td>GGCGGCGTGGTGCTGCACTTCGACGTGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTCGAAGTGCAGCACCACGCCGCCCCAGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>D9T</td>\n",
       "      <td>F</td>\n",
       "      <td>GACTGCTGATCAGCACCAGCTACGACCCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GGTCGTAGCTGGTGCTGATCAGCAGTCTCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>L161H</td>\n",
       "      <td>F</td>\n",
       "      <td>GACCTGAGCAGACACGCCAACTACCTGAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTTCAGGTAGTTGGCGTGTCTGCTCAGGTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>F281K</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A48E</td>\n",
       "      <td>F</td>\n",
       "      <td>GTGATCGGCAGAGAGCAGAACCCCTGGAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CCAGGGGTTCTGCTCTCTGCCGATCACCACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>L161R</td>\n",
       "      <td>F</td>\n",
       "      <td>CGACCTGAGCAGAAGGGCCAACTACCTGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>TCAGGTAGTTGGCCCTTCTGCTCAGGTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>E265I</td>\n",
       "      <td>F</td>\n",
       "      <td>CTGGGGCGGCGTGATCCTGCACTTCGACGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTCGAAGTGCAGGATCACGCCGCCCCAGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>F147S</td>\n",
       "      <td>F</td>\n",
       "      <td>CAAGGACAGAGGCAGCCACCACGGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GGTGCCGTGGTGGCTGCCTCTGTCCTTGGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>L161Q</td>\n",
       "      <td>F</td>\n",
       "      <td>CGACCTGAGCAGACAGGCCAACTACCTGAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTTCAGGTAGTTGGCCTGTCTGCTCAGGTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>L161M</td>\n",
       "      <td>F</td>\n",
       "      <td>CGACCTGAGCAGAATGGCCAACTACCTGAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GGTAGTTGGCCATTCTGCTCAGGTCGGCGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A48R</td>\n",
       "      <td>F</td>\n",
       "      <td>GTGATCGGCAGAAGACAGAACCCCTGGAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CCAGGGGTTCTGTCTTCTGCCGATCACCACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>L161Y</td>\n",
       "      <td>F</td>\n",
       "      <td>GACCTGAGCAGATACGCCAACTACCTGAACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTTCAGGTAGTTGGCGTATCTGCTCAGGTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>A48K</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>F147R</td>\n",
       "      <td>F</td>\n",
       "      <td>CAAGGACAGAGGCAGACACCACGGCACCCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GGTGCCGTGGTGTCTGCCTCTGTCCTTGGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>P229E</td>\n",
       "      <td>F</td>\n",
       "      <td>GACCCCCGACCTGGAGAACTTCGCCGAGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CTCGGCGAAGTTCTCCAGGTCGGGGGTCTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>E265T</td>\n",
       "      <td>F</td>\n",
       "      <td>GGGCGGCGTGACCCTGCACTTCGACGTGGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTCGAAGTGCAGGGTCACGCCGCCCCAGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>C311F</td>\n",
       "      <td>F</td>\n",
       "      <td>GCTGCAGCAGGAGTTCGAGGCCCTGCTGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>CAGCAGGGCCTCGAACTCCTGCTGCAGCATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Q26T</td>\n",
       "      <td>F</td>\n",
       "      <td>GTGCATCTTCAGAACCATGCCCGCCACCCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GGGTGGCGGGCATGGTTCTGAAGATGCACTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>E265R</td>\n",
       "      <td>F</td>\n",
       "      <td>CTGGGGCGGCGTGAGACTGCACTTCGACGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GTCGAAGTGCAGTCTCACGCCGCCCCAGGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>E265K</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>F281L</td>\n",
       "      <td>F</td>\n",
       "      <td>CAGAGCCCAGGTGCTGACCGACAGCCTGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>GCTGTCGGTCAGCACCTGGGCTCTGGTGATG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Target_Mutation Orientation                          Sequence\n",
       "0            Y139V           F   GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC\n",
       "1              NaN           R   CTTGGTCTCTCTCACGGCGCTGCCGCTCACC\n",
       "2            Y139T           F   GAGCGGCAGCGCCACCAGAGAGACCAAGGAC\n",
       "3              NaN           R   CTTGGTCTCTCTGGTGGCGCTGCCGCTCACC\n",
       "4            Y139S           F   GAGCGGCAGCGCCAGCAGAGAGACCAAGGAC\n",
       "5              NaN           R   CTTGGTCTCTCTGCTGGCGCTGCCGCTCACC\n",
       "6             A48D           F                               NaN\n",
       "7              NaN           R                               NaN\n",
       "8            E265V           F    GGCGGCGTGGTGCTGCACTTCGACGTGGAG\n",
       "9              NaN           R   GTCGAAGTGCAGCACCACGCCGCCCCAGGTG\n",
       "10             D9T           F   GACTGCTGATCAGCACCAGCTACGACCCCTG\n",
       "11             NaN           R   GGTCGTAGCTGGTGCTGATCAGCAGTCTCAG\n",
       "12           L161H           F   GACCTGAGCAGACACGCCAACTACCTGAACC\n",
       "13             NaN           R   GTTCAGGTAGTTGGCGTGTCTGCTCAGGTCG\n",
       "14           F281K           F                               NaN\n",
       "15             NaN           R                               NaN\n",
       "16            A48E           F   GTGATCGGCAGAGAGCAGAACCCCTGGAAGG\n",
       "17             NaN           R   CCAGGGGTTCTGCTCTCTGCCGATCACCACG\n",
       "18           L161R           F   CGACCTGAGCAGAAGGGCCAACTACCTGAAC\n",
       "19             NaN           R    TCAGGTAGTTGGCCCTTCTGCTCAGGTCGG\n",
       "20           E265I           F   CTGGGGCGGCGTGATCCTGCACTTCGACGTG\n",
       "21             NaN           R   GTCGAAGTGCAGGATCACGCCGCCCCAGGTG\n",
       "22           F147S           F   CAAGGACAGAGGCAGCCACCACGGCACCCTG\n",
       "23             NaN           R   GGTGCCGTGGTGGCTGCCTCTGTCCTTGGTC\n",
       "24           L161Q           F  CGACCTGAGCAGACAGGCCAACTACCTGAACC\n",
       "25             NaN           R  GTTCAGGTAGTTGGCCTGTCTGCTCAGGTCGG\n",
       "26           L161M           F  CGACCTGAGCAGAATGGCCAACTACCTGAACC\n",
       "27             NaN           R  GGTAGTTGGCCATTCTGCTCAGGTCGGCGTTC\n",
       "28            A48R           F   GTGATCGGCAGAAGACAGAACCCCTGGAAGG\n",
       "29             NaN           R   CCAGGGGTTCTGTCTTCTGCCGATCACCACG\n",
       "30           L161Y           F   GACCTGAGCAGATACGCCAACTACCTGAACC\n",
       "31             NaN           R  GTTCAGGTAGTTGGCGTATCTGCTCAGGTCGG\n",
       "32            A48K           F                               NaN\n",
       "33             NaN           R                               NaN\n",
       "34           F147R           F   CAAGGACAGAGGCAGACACCACGGCACCCTG\n",
       "35             NaN           R   GGTGCCGTGGTGTCTGCCTCTGTCCTTGGTC\n",
       "36           P229E           F   GACCCCCGACCTGGAGAACTTCGCCGAGACC\n",
       "37             NaN           R   CTCGGCGAAGTTCTCCAGGTCGGGGGTCTTG\n",
       "38           E265T           F   GGGCGGCGTGACCCTGCACTTCGACGTGGAG\n",
       "39             NaN           R   GTCGAAGTGCAGGGTCACGCCGCCCCAGGTG\n",
       "40           C311F           F   GCTGCAGCAGGAGTTCGAGGCCCTGCTGGTG\n",
       "41             NaN           R   CAGCAGGGCCTCGAACTCCTGCTGCAGCATG\n",
       "42            Q26T           F   GTGCATCTTCAGAACCATGCCCGCCACCCAG\n",
       "43             NaN           R   GGGTGGCGGGCATGGTTCTGAAGATGCACTC\n",
       "44           E265R           F   CTGGGGCGGCGTGAGACTGCACTTCGACGTG\n",
       "45             NaN           R   GTCGAAGTGCAGTCTCACGCCGCCCCAGGTG\n",
       "46           E265K           F                               NaN\n",
       "47             NaN           R                               NaN\n",
       "48           F281L           F   CAGAGCCCAGGTGCTGACCGACAGCCTGAAC\n",
       "49             NaN           R   GCTGTCGGTCAGCACCTGGGCTCTGGTGATG"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Alignment(seqA='AGCACCCTGAGACTGCTGATCAGCGACAGCTACGACCCCTGGTTCAACCTGGCCGTGGAGGAGTGCATCTTCAGACAGATGCCCGCCACCCAGAGAGTGCTGTTCCTGGTGAGAAACGCCGACACCGTGGTGATCGGCAGAGCCCAGAACCCCTGGAAGGAGTGCAACACCAGAAGAATGGAGGAGGACAACGTGAGACTGGCCAGAAGAAGCAGCGGCGGCGGCGCCGTGTTCCACGACCTGGGCAACACCTGCTTCACCTTCATGGCCGGCAAGCCCGAGTACGACAAGACCATCAGCACCAGCATCGTGCTGAACGCCCTGAACGCCCTGGGCGTGAGCGCCGAGGCCAGCGGCAGAAACGACCTGGTGGTGAAGACCGTGGAGGGCGACAGAAAGGTGAGCGGCAGCGCCTACAGAGAGACCAAGGACAGAGGCTTCCACCACGGCACCCTGCTGCTGAACGCCGACCTGAGCAGACTGGCCAACTACCTGAACCCCGACAAGAAGAAGCTGGCCGCCAAGGGCATCACCAGCGTGAGAAGCAGAGTGACCAACCTGACCGAGCTGCTGCCCGGCATCACCCACGAGCAGGTGTGCGAGGCCATCACCGAGGCCTTCTTCGCCCACTACGGCGAGAGAGTGGAGGCCGAGATCATCAGCCCCAACAAGACCCCCGACCTGCCCAACTTCGCCGAGACCTTCGCCAGACAGAGCAGCTGGGAGTGGAACTTCGGCCAGGCCCCCGCCTTCAGCCACCTGCTGGACGAGAGATTCACCTGGGGCGGCGTGGAGCTGCACTTCGACGTGGAGAAGGGCCACATCACCAGAGCCCAGGTGTTCACCGACAGCCTGAACCCCGCCCCCCTGGAGGCCCTGGCCGGCAGACTGCAGGGCTGCCTGTACAGAGCCGACATGCTGCAGCAGGAGTGCGAGGCCCTGCTGGTGGACTTCCCCGAGCAGGAGAAGGAGCTGAGAGAGCTGAGCGCCTGGATGGCCGGCGCCGTGAGATAAgaattctgcagatatccatcacactgg', seqB='-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------GAGCGGCAGCGCCGTGAGAGAGACCAAGGAC---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------', score=56.0, start=0, end=1041)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hhh\n",
      "hhh\n"
     ]
    }
   ],
   "source": [
    "translate('hhhhhh', dna_codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': 'F',\n",
       " 'CTT': 'L',\n",
       " 'ATT': 'I',\n",
       " 'GTT': 'V',\n",
       " 'TTC': 'F',\n",
       " 'CTC': 'L',\n",
       " 'ATC': 'I',\n",
       " 'GTC': 'V',\n",
       " 'TTA': 'L',\n",
       " 'CTA': 'L',\n",
       " 'ATA': 'I',\n",
       " 'GTA': 'V',\n",
       " 'TTG': 'L',\n",
       " 'CTG': 'L',\n",
       " 'ATG': 'M',\n",
       " 'GTG': 'V',\n",
       " 'TCT': 'S',\n",
       " 'CCT': 'P',\n",
       " 'ACT': 'T',\n",
       " 'GCT': 'A',\n",
       " 'TCC': 'S',\n",
       " 'CCC': 'P',\n",
       " 'ACC': 'T',\n",
       " 'GCC': 'A',\n",
       " 'TCA': 'S',\n",
       " 'CCA': 'P',\n",
       " 'ACA': 'T',\n",
       " 'GCA': 'A',\n",
       " 'TCG': 'S',\n",
       " 'CCG': 'P',\n",
       " 'ACG': 'T',\n",
       " 'GCG': 'A',\n",
       " 'TAT': 'Y',\n",
       " 'CAT': 'H',\n",
       " 'AAT': 'N',\n",
       " 'GAT': 'D',\n",
       " 'TAC': 'Y',\n",
       " 'CAC': 'H',\n",
       " 'AAC': 'N',\n",
       " 'GAC': 'D',\n",
       " 'TAA': 'STOP',\n",
       " 'CAA': 'Q',\n",
       " 'AAA': 'K',\n",
       " 'GAA': 'E',\n",
       " 'TAG': 'STOP',\n",
       " 'CAG': 'Q',\n",
       " 'AAG': 'K',\n",
       " 'GAG': 'E',\n",
       " 'TGT': 'C',\n",
       " 'CGT': 'R',\n",
       " 'AGT': 'S',\n",
       " 'GGT': 'G',\n",
       " 'TGC': 'C',\n",
       " 'CGC': 'R',\n",
       " 'AGC': 'S',\n",
       " 'GGC': 'G',\n",
       " 'TGA': 'STOP',\n",
       " 'CGA': 'R',\n",
       " 'AGA': 'R',\n",
       " 'GGA': 'G',\n",
       " 'TGG': 'W',\n",
       " 'CGG': 'R',\n",
       " 'AGG': 'R',\n",
       " 'GGG': 'G'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'template' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 11\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcodon_table\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m dna_codons\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mcolors\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m rasmol\n\u001b[1;32m     10\u001b[0m alignments \u001b[38;5;241m=\u001b[39m pairwise2\u001b[38;5;241m.\u001b[39malign\u001b[38;5;241m.\u001b[39mglobalms(\n\u001b[0;32m---> 11\u001b[0m     template, subject, \u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m10\u001b[39m, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m, penalize_end_gaps\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m     12\u001b[0m )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'template' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "alignments = pairwise2.align.globalms(\n",
    "    template, subject, 2, 0, -10, -1, penalize_end_gaps=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Something has gone wrong :(",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 46\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m     45\u001b[0m     template_limits \u001b[39m=\u001b[39m (\u001b[39m5000\u001b[39m, \u001b[39m5740\u001b[39m)\n\u001b[0;32m---> 46\u001b[0m fig, stats \u001b[39m=\u001b[39m visualize(\u001b[39mid\u001b[39m, abidata, template\u001b[39m=\u001b[39mtemplate, strand\u001b[39m=\u001b[39mstrand, fig\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, region\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mread\u001b[39m\u001b[39m\"\u001b[39m, translation_limits \u001b[39m=\u001b[39m (\u001b[39m4702\u001b[39m, \u001b[39mNone\u001b[39;00m), template_limits \u001b[39m=\u001b[39m template_limits)\n\u001b[1;32m     47\u001b[0m fig\u001b[39m.\u001b[39msavefig(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mid\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m.png\u001b[39m\u001b[39m'\u001b[39m, bbox_inches \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mtight\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     49\u001b[0m stat_list\u001b[39m.\u001b[39mappend(stats)\n",
      "File \u001b[0;32m~/lab/Andrew/pysanger.py:403\u001b[0m, in \u001b[0;36mvisualize\u001b[0;34m(title, abidata, template, strand, fig, region, translation_limits, template_limits)\u001b[0m\n\u001b[1;32m    401\u001b[0m \u001b[39mif\u001b[39;00m n_template_nt \u001b[39m!=\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m    402\u001b[0m     \u001b[39mprint\u001b[39m(n_template_nt)\n\u001b[0;32m--> 403\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m'\u001b[39m\u001b[39mSomething has gone wrong :(\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m    405\u001b[0m insertions \u001b[39m=\u001b[39m []\n\u001b[1;32m    406\u001b[0m deletions \u001b[39m=\u001b[39m []\n",
      "\u001b[0;31mValueError\u001b[0m: Something has gone wrong :("
     ]
    }
   ],
   "source": [
    "from pysanger import *\n",
    "\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "def tokenize_fp_make_id(inp):\n",
    "    tokens = inp.replace('-', '_').split('_')\n",
    "    user = tokens[0]\n",
    "    construct = tokens[2]\n",
    "    num = tokens[3]\n",
    "\n",
    "    forward = 'Mcherry' in tokens\n",
    "    reverse = 'WPRE' in tokens\n",
    "\n",
    "    if not forward and not reverse:\n",
    "        raise Exception('Neither forward nor reverse! Invalid filename.')\n",
    "    else:\n",
    "        if forward:\n",
    "            primer = 'MCHERRY'\n",
    "            direction = 'FORWARD'\n",
    "            strand = 1\n",
    "        else:\n",
    "            primer = 'WPRE'\n",
    "            direction = 'REVERSE'\n",
    "            strand = -1\n",
    "\n",
    "    id = f\"{user}_{construct}_{num}_{primer}_{direction}\"\n",
    "\n",
    "    return id, strand\n",
    "\n",
    "#read in template plasmid\n",
    "with open(\"data/bfp-expression-control-mcherry-flag-hlpla-pcdna3-w37v.fasta\") as file:\n",
    "    for record in SeqIO.parse(file, \"fasta\"):\n",
    "        template = record.seq\n",
    "\n",
    "fp = 'PC_812288-3088_48_PC-442--Mcherry-Seq-Forward_H11.ab1'\n",
    "id, strand = tokenize_fp_make_id(fp)\n",
    "abidata = abi_to_dict(f\"peter_data/{fp}\")\n",
    "\n",
    "if strand == 1:\n",
    "    template_limits = (4660, 5250)\n",
    "else:\n",
    "    template_limits = (5000, 5740)\n",
    "fig, stats = visualize(id, abidata, template=template, strand=strand, fig=None, region=\"read\", translation_limits = (4702, None), template_limits = template_limits)\n",
    "fig.savefig(f'{id}.png', bbox_inches = \"tight\")\n",
    "\n",
    "stat_list.append(stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>alignment_score</th>\n",
       "      <th>nt_insertions</th>\n",
       "      <th>nt_deletions</th>\n",
       "      <th>nt_mismatches</th>\n",
       "      <th>silent_mutations</th>\n",
       "      <th>missense_mutations</th>\n",
       "      <th>gap_mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PC_574_2_MCHERRY_FORWARD</td>\n",
       "      <td>1102.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4792(C-&gt;G)</td>\n",
       "      <td>31</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PC_574_2_MCHERRY_FORWARD</td>\n",
       "      <td>1102.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4792(C-&gt;G)</td>\n",
       "      <td>31</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         id  alignment_score nt_insertions nt_deletions  \\\n",
       "0  PC_574_2_MCHERRY_FORWARD           1102.0                              \n",
       "0  PC_574_2_MCHERRY_FORWARD           1102.0                              \n",
       "\n",
       "  nt_mismatches silent_mutations missense_mutations gap_mutations  \n",
       "0    4792(C->G)               31                                   \n",
       "0    4792(C->G)               31                                   "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([stats, stats])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>alignment_score</th>\n",
       "      <th>nt_insertions</th>\n",
       "      <th>nt_deletions</th>\n",
       "      <th>nt_mismatches</th>\n",
       "      <th>silent_mutations</th>\n",
       "      <th>missense_mutations</th>\n",
       "      <th>gap_mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;built-in function id&gt;</td>\n",
       "      <td>1102.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4792(C-&gt;G)</td>\n",
       "      <td>31</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       id  alignment_score nt_insertions nt_deletions  \\\n",
       "0  <built-in function id>           1102.0                              \n",
       "\n",
       "  nt_mismatches silent_mutations missense_mutations gap_mutations  \n",
       "0    4792(C->G)               31                                   "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
